# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [ ]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [6]:
x.shape

(506, 13)

In [8]:
y.shape

(506,)

In [7]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

In [96]:
y[:5]

array([24. , 21.6, 34.7, 33.4, 36.2])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the random 60% datapoints from the whole data, make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [9]:
import random


In [10]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    selected_rows = np.random.choice(len(input_data), 303, replace = False)
    selected_rows = np.sort(selected_rows)


    replaced_rows = np.random.choice(selected_rows, 203, replace = False)
    replaced_rows = np.sort(replaced_rows)

    selected_columns = np.sort(np.random.choice(input_data.shape[1], size=random.randint(3, input_data.shape[1]), replace =False))
    
    sample_data = input_data[selected_rows[:,None],selected_columns]
    target_sameple_data = target_data[selected_rows]

    replicating_data = input_data[replaced_rows[:,None],selected_columns]
    target_replicating_data = target_data[replaced_rows]

    sampled_input_data = np.vstack((sample_data, replicating_data))
    sampled_target_data = np.vstack((target_sameple_data.reshape(-1,1),target_replicating_data.reshape(-1,1)))

    return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists
    

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [11]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [53]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

In [54]:
for i in range(0,30):
  sampled_input_data, sampled_target_data, selected_rows, selected_columns = generating_samples(x,y)
  #grader_samples(final_sample_data , final_target_data,selected_rows,selected_columns)
  # Append the info in the respective list.
  list_input_data.append(sampled_input_data)
  list_output_data.append(sampled_target_data)
  list_selected_row.append(selected_rows)
  list_selected_columns.append(selected_columns)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [55]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [56]:
from sklearn.tree import DecisionTreeRegressor 
list_of_all_models = []
for i in range(0, 30):
  input_data = list_input_data[i]
  target_data = list_output_data[i]
  classifier = DecisionTreeRegressor(max_depth=None,min_samples_split=2) 
  classifier.fit(input_data, target_data) 
  list_of_all_models.append(classifier)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [60]:
x[:]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [57]:
import statistics

list_y_predict = []
for i in range(0,30):
  pred_y = list_of_all_models[i].predict(x[:,list_selected_columns[i]]) 
  list_y_predict.append(pred_y)

In [62]:
len(list_y_predict[0])

506

In [64]:
list_y_predict[0]

array([24. , 21.6, 34.7, 48.5, 36.2, 28.7, 20.1, 27.1, 16.5, 18.9, 15. ,
       23.1, 21.7, 19.6, 18.2, 24.3, 20.6, 19.6, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 20.2, 12.7, 19.6, 13.2,
       13.1, 13.1, 18.9, 20. , 21. , 24.7, 30.8, 44. , 31.1, 25.3, 24.7,
       21.2, 19.3, 19.3, 16.6, 18.7, 19.4, 22.5, 20.5, 25. , 19.8, 18.9,
       27.9, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 23.9, 25. , 28.7, 23.5,
       24. , 18.2, 17.4, 20.9, 22.6, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20. , 21.2, 23.7, 28. , 23.9, 23.2, 22.9, 18.7, 26.6, 22.5, 22.2,
       23.6, 28.7, 19.8, 22. , 22.9, 28.7, 19.5, 28.4, 21.4, 44.8, 43.8,
       33.2, 23.7, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 20.1, 19.4,
       21.7, 22.8, 23.1, 18.7, 19.2, 17.6, 21.2, 19.2, 20.4, 19.3, 29.6,
       21.4, 17.3, 17.3, 17.3, 21.4, 17.8, 16.2, 18. , 15.6, 19.2, 23.8,
       18.2, 18. , 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 13.3, 13.3, 27.1,
       27.1, 27.1, 17.8, 17.8, 14.6, 17.8, 17.8, 19

In [ ]:
# Calculating median for each data point from predicted y of each model
final_y_predict = [] 

for i in range(0, len(x)):
  median_y = []
  for j in range(0, 30):
    median_y.append(list_y_predict[j][i]) 
  median_ = statistics.median(median_y) 
  final_y_predict.append(median_)

In [59]:
# Calculating MSE :
from sklearn.metrics import mean_squared_error 
mse = mean_squared_error(y, final_y_predict) 
print("=" * 50)
print("MSE on train data set : " , mse) 
print("=" * 50)

MSE on train data set :  0.0338586956521739


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [74]:
y_sample_predict = []
pred_y = list_of_all_models[1].predict(x[0][list_selected_columns[1]].reshape(1,-1)) 
y_sample_predict.append(pred_y[0])

In [82]:
x[0][list_selected_columns[1]].shape

(11,)

In [81]:
x[0][list_selected_columns[1]].reshape(1,-1).shape

(1, 11)

In [80]:
x[0].shape

(13,)

In [85]:
pred_y = list_of_all_models[j].predict(x[0][list_selected_columns[j]].reshape(1,-1)) 
pred_y[0]

24.0

In [86]:
oob_y_predict = []
y_sample_predict = []
# iterating through each data point. 
for i in range (0,506):
  data = x[i]
  for j in range(0,30):#30
    if i not in list_selected_row[j]:
      pred_y = list_of_all_models[j].predict(data[list_selected_columns[j]].reshape(1,-1)) 
      y_sample_predict.append(pred_y[0])
  y_sample_predict_oobs = statistics.median(y_sample_predict) 
  y_sample_predict.clear() 
  oob_y_predict.append(y_sample_predict_oobs)

#print((oob_y_predict))
#Calculate OObs score now :
total_error = 0
for i in range(0, 506):
  error = y[i] - oob_y_predict[i] 
  error = error * error #Square error :
  total_error = total_error + error 
oobs_score = total_error/506 
print("=" * 35)
print("OOBS Score : " , oobs_score) 
print("=" * 35)

OOBS Score :  11.253729518006153


# <font color='blue'><b>Task 2</b></font>

In [89]:
# repeat Task 1 35 times :
train_mse_list = [] 
oobs_score_list = []
for iteration in range (1 , 36):
    list_input_data =[] 
    list_output_data =[] 
    list_selected_row= [] 
    list_selected_columns=[] 
    list_of_all_models = [] 
    list_predict = [] 
    oob_y_predict = [] 
    y_sample_predict = []
  
    for i in range(0,30):
        final_sample_data, final_target_data, selected_rows, selected_columns = generating_samples(x,y)
        # Append the info in the respective list.
        list_input_data.append(final_sample_data) 
        list_output_data.append(final_target_data) 
        list_selected_row.append(selected_rows) 
        list_selected_columns.append(selected_columns)

    #Write code for building regression trees
    for i in range(0, 30):
        input_data = list_input_data[i]
        target_data = list_output_data[i]
        classifier = DecisionTreeRegressor(max_depth=None,min_samples_split=2) 
        classifier.fit(input_data,target_data) 
        list_of_all_models.append(classifier)
    
    # write code for Calculating MSE :
    for i in range(0,30):
        pred_y = list_of_all_models[i].predict(x[:,list_selected_columns[i]]) 
        list_predict.append(pred_y)
      # Calculating median for each data point from predicted y of each model.
    final_y_predict = [] 
    for i in range(0 , 506):
        med_y = []
        for j in range(0 , 30):
            med_y.append(list_predict[j][i]) 
        med = statistics.median(med_y) 
        final_y_predict.append(med)

    # Calculating MSE :
    mse = mean_squared_error(y, final_y_predict) 
    print("=" * 30)
    print("Iteration: " , iteration) 
    print("MSE on train data =" , round(mse,3)) 
    train_mse_list.append(mse)
    # Calculating OOBS score
    # iterating through each data point. 
    for i in range (0,506):
        data = x[i]
        for j in range(0,30):
            if i not in list_selected_row[j]:
                pred_y = list_of_all_models[j].predict(data[list_selected_columns[j]].reshape(1,-1)) 
                y_sample_predict.append(pred_y[0])
        y_sample_predict_oobs = statistics.median(y_sample_predict) 
        y_sample_predict.clear() 
        oob_y_predict.append(y_sample_predict_oobs)

    #Calculate OObs score now :
    total_error = 0
    for i in range(0,506):
      error = y[i] - oob_y_predict[i] #Square error :
      error = error * error 
      total_error += error
    oobs_score = total_error/506 
    print("OOBS Score = " , round(oobs_score,3)) 
    print("=" * 30) 
    oobs_score_list.append(oobs_score)

Iteration:  1
MSE on train data = 0.196
OOBS Score =  14.501
Iteration:  2
MSE on train data = 0.086
OOBS Score =  16.666
Iteration:  3
MSE on train data = 0.241
OOBS Score =  13.803
Iteration:  4
MSE on train data = 0.01
OOBS Score =  14.063
Iteration:  5
MSE on train data = 0.076
OOBS Score =  14.177
Iteration:  6
MSE on train data = 0.029
OOBS Score =  11.661
Iteration:  7
MSE on train data = 0.055
OOBS Score =  14.547
Iteration:  8
MSE on train data = 0.066
OOBS Score =  12.253
Iteration:  9
MSE on train data = 0.051
OOBS Score =  13.699
Iteration:  10
MSE on train data = 0.016
OOBS Score =  13.21
Iteration:  11
MSE on train data = 0.062
OOBS Score =  12.837
Iteration:  12
MSE on train data = 0.267
OOBS Score =  15.574
Iteration:  13
MSE on train data = 0.028
OOBS Score =  15.608
Iteration:  14
MSE on train data = 0.076
OOBS Score =  14.805
Iteration:  15
MSE on train data = 0.069
OOBS Score =  16.16
Iteration:  16
MSE on train data = 0.042
OOBS Score =  11.722
Iteration:  17
MSE o

### Compute Confidence Interval

![alt text](https://datatab.net/assets/tutorial/Confidence_interval_formula.png)

In [94]:
#Converting list to array
import math
train_mse = np.asarray(train_mse_list) 
oobs_score = np.asarray(oobs_score_list)

# Compute mean
mean_mse = np.mean(train_mse) 
mean_oobs = np.mean(oobs_score)

#Compute Standard Deviation:
std_mse = np.std(train_mse) 
std_oobs = np.std(oobs_score)

#Compute Standard error:
sqrt_n = math.sqrt(30) 
standard_error_mse = std_mse/sqrt_n 
standard_error_obs = std_oobs/sqrt_n

#CI for MSE:
lower_limit_mse = mean_mse - 2 * (standard_error_mse) 
upper_limit_mse = mean_mse + 2 * (standard_error_mse)
print("=" * 60)
print("CI for MSE :" , "[", lower_limit_mse , "," , upper_limit_mse,"]")

#CI for MSE :
lower_limit_oobs = mean_oobs - 2 * (standard_error_obs) 
upper_limit_oobs = mean_oobs + 2 * (standard_error_obs)
print("CI for oob :" , "[", lower_limit_oobs , "," , upper_limit_oobs,"]" ) 
print("=" * 60)

CI for MSE : [ 0.05100328948561962 , 0.14635609957901446 ]
CI for oob : [ 13.328001251278668 , 14.699926080114903 ]


## Observation / Interpretation of above Confidence Interval
- By definition we know the interpretation of a 95% confidence interval for the population mean as - If repeated random samples were taken and the 95% confidence interval was computed for each sample, 95% of the intervals would contain the population mean.

So in this case

* MSE - There is a 95% chance that the confidence interval of (0.05100328948561962, 0.14635609957901446) contains the true population mean of MSE.
* OOB Score - There is a 95% chance that the confidence interval of (13.328001251278668, 14.699926080114903) contains the true population mean of OOB Score.


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [95]:
x_q = np.array([0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60])
total_price = 0
for m in range(0,30):
  pred_y = list_of_all_models[m].predict(x_q[list_selected_columns[m]].reshape(1,-1))
  total_price += pred_y

# Not predicted price for the query point will be average of all prices predicted.
predicted_price = total_price/30
print("=" * 50)
print("Predicted price for query point x_q is :" , predicted_price) 
print("=" * 50)

Predicted price for query point x_q is : [19.33333333]


<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

# Observation from Task-1:
- The Mean Squared Error (MSE) of 0.0338586956521739 and the Out-of-Bag (OOB) score of 11.253729518006153 are evaluation metrics used to assess the performance of a Random Forest model applied on the Boston housing dataset.

- A lower MSE value indicates that the model is making fewer errors in predicting the target variable (in this case, housing prices). A MSE value of 0.0338 suggests that the model's predictions are relatively accurate.

- The OOB score is an estimate of the model's performance on unseen data (i.e., data not used in training). An OOB score of 11.25 indicates that, on average, the model's predictions are off by about 11.25 units of the target variable.

- In general, these values suggest that the Random Forest model is making accurate predictions and that the model's performance is good. However, these metrics should be compared with other evaluation metrics (such as R2 score) and the results of models trained with other algorithms to get a more comprehensive view of the model's performance.

## Observation from Task-2:

- The Confidence Interval (CI) of the Mean Squared Error (MSE) [0.05100328948561962, 0.14635609957901446] and Out-of-Bag (OOB) score [13.328001251278668, 14.699926080114903] provides a range of values in which the true values of the metrics are likely to lie, given a certain level of confidence.

- The CI of the MSE provides a range of values for the accuracy of the model's predictions. The interval [0.05100328948561962, 0.14635609957901446] suggests that with a certain level of confidence, the true MSE of the model lies between 0.05 and 0.15. A smaller MSE indicates more accurate predictions, so a CI that includes lower MSE values is preferred.

- The CI of the OOB score provides a range of values for the performance of the model on unseen data. The interval [13.328001251278668, 14.699926080114903] suggests that with a certain level of confidence, the true OOB score of the model lies between 13.33 and 14.70. A lower OOB score indicates better performance, so a CI that includes lower OOB scores is preferred.

- Overall, these values suggest that the model's performance on the Boston housing dataset is within the specified range of confidence. However, it is important to keep in mind that the intervals may be wide, indicating high variability or low confidence in the estimated values.

# Observation from Task-3:
*(target variable(medv): median value of owner-occupied homes in \$1000s.)*
- The prediction value of 19.33333333 for the query point on the Boston Housing dataset using a Random Forest model suggests that the model is predicting a median housing price of approximately $19,333.33 (19.333.*$1000). However, it is important to note that this value is just one prediction and may not accurately represent the actual median housing price in the area.

- In order to better understand the prediction, it would be necessary to examine other factors such as the accuracy of the model, the quality of the data used for training, and the specific features used for prediction. Additionally, it may be helpful to compare this prediction to actual median housing prices in the area to determine the accuracy of the prediction.


"""
# **References:**

The following links/blogs/resources are references which i referred While writing above code
- https://www.kaggle.com/code/paulrohan2020/implementation-of-bootstrap-in-random-forest/notebook

- https://github.com/tufts-ml-courses/comp135-20f-assignments/blob/master/labs/day15_Bootstrap_and_RandomForest.ipynb
- https://github.com/selva86/datasets/blob/master/BostonHousing.csv
- https://stats.stackexchange.com/questions/354336/what-happens-when-bootstrapping-isnt-used-in-sklearn-randomforestclassifier 
- ChatGPT
- stackoverflow
- Quora
- Yotube

"""